docs/configuring_tor.md
### Table of Contents
1. [Overview](#Overview)<br />
2. [Client-Only](#Client)<br />
2.1 [Description](#ClientDescription)<br />
2.2 [Command Line Example](#ClientCLIExample)<br />
2.3 [Config File Example](#ClientConfigFileExample)<br />
3. [Client-Server via Tor Hidden Service](#HiddenService)<br />
3.1 [Description](#HiddenServiceDescription)<br />
3.2 [Command Line Example](#HiddenServiceCLIExample)<br />
3.3 [Config File Example](#HiddenServiceConfigFileExample)<br />
4. [Bridge Mode (Not Anonymous)](#Bridge)<br />
4.1 [Description](#BridgeDescription)<br />
4.2 [Command Line Example](#BridgeCLIExample)<br />
4.3 [Config File Example](#BridgeConfigFileExample)<br />
5. [Tor Stream Isolation](#TorStreamIsolation)<br />
5.1 [Description](#TorStreamIsolationDescription)<br />
5.2 [Command Line Example](#TorStreamIsolationCLIExample)<br />
5.3 [Config File Example](#TorStreamIsolationFileExample)<br />

<a name="Overview" />

### 1. Overview

hcd provides full support for anonymous networking via the
[Tor Project](https://www.torproject.org/), including [client-only](#Client)
and [hidden service](#HiddenService) configurations along with
[stream isolation](#TorStreamIsolation).  In addition, hcd supports a hybrid,
[bridge mode](#Bridge) which is not anonymous, but allows it to operate as a
bridge between regular nodes and hidden service nodes without routing the
regular connections through Tor.

While it is easier to only run as a client, it is more beneficial to the Hc
network to run as both a client and a server so others may connect to you to as
you are connecting to them.  We recommend you take the time to setup a Tor
hidden service for this reason.

<a name="Client" />

### 2. Client-Only

<a name="ClientDescription" />

**2.1 Description**<br />

Configuring hcd as a Tor client is straightforward.  The first step is
obviously to install Tor and ensure it is working. Once that is done, all that
typically needs to be done is to specify the `--proxy` flag via the hcd command
line or in the hcd configuration file.  Typically the Tor proxy address will be
127.0.0.1:9050 (if using standalone Tor) or 127.0.0.1:9150 (if using the Tor
Browser Bundle).  If you have Tor configured to require a username and password,
you may specify them with the `--proxyuser` and `--proxypass` flags.

By default, hcd assumes the proxy specified with `--proxy` is a Tor proxy and
hence will send all traffic, including DNS resolution requests, via the
specified proxy.

NOTE: Specifying the `--proxy` flag disables listening by default since you will
not be reachable for inbound connections unless you also configure a Tor
[hidden service](#HiddenService).

<a name="ClientCLIExample" />

**2.2 Command Line Example**<br />

```bash
$ ./hcd --proxy=127.0.0.1:9050
```

<a name="ClientConfigFileExample" />

**2.3 Config File Example**<br />

```text
[Application Options]

proxy=127.0.0.1:9050
```

<a name="HiddenService" />

### 3. Client-Server via Tor Hidden Service

<a name="HiddenServiceDescription" />

**3.1 Description**<br />

The first step is to configure Tor to provide a hidden service.  Documentation
for this can be found on the Tor project website
[here](https://www.torproject.org/docs/tor-hidden-service.html.en).  However,
there is no need to install a web server locally as the linked instructions
discuss since hcd will act as the server.

In short, the instructions linked above entail modifying your `torrc` file to
add something similar to the following, restarting Tor, and opening the
`hostname` file in the `HiddenServiceDir` to obtain your hidden service .onion
address.

```text
HiddenServiceDir /var/tor/hcd
HiddenServicePort 9108 127.0.0.1:9108
```

Once Tor is configured to provide the hidden service and you have obtained your
generated .onion address, configuring hcd as a Tor hidden service requires
three flags:
* `--proxy` to identify the Tor (SOCKS 5) proxy to use for outgoing traffic.
  This is typically 127.0.0.1:9050.
* `--listen` to enable listening for inbound connections since `--proxy`
  disables listening by default
* `--externalip` to set the .onion address that is advertised to other peers

<a name="HiddenServiceCLIExample" />

**3.2 Command Line Example**<br />

```bash
$ ./hcd --proxy=127.0.0.1:9050 --listen=127.0.0.1 --externalip=fooanon.onion
```

<a name="HiddenServiceConfigFileExample" />

**3.3 Config File Example**<br />

```text
[Application Options]

proxy=127.0.0.1:9050
listen=127.0.0.1
externalip=fooanon.onion
```

<a name="Bridge" />

### 4. Bridge Mode (Not Anonymous)

<a name="BridgeDescription" />

**4.1 Description**<br />

hcd provides support for operating as a bridge between regular nodes and hidden
service nodes.  In particular this means only traffic which is directed to or
from a .onion address is sent through Tor while other traffic is sent normally.
_As a result, this mode is **NOT** anonymous._

This mode works by specifying an onion-specific proxy, which is pointed at Tor,
by using the `--onion` flag via the hcd command line or in the hcd
configuration file.  If you have Tor configured to require a username and
password, you may specify them with the `--onionuser` and `--onionpass` flags.

NOTE: This mode will also work in conjunction with a hidden service which means
you could accept inbound connections both via the normal network and to your
hidden service through the Tor network.  To enable your hidden service in bridge
mode, you only need to specify your hidden service's .onion address via the
`--externalip` flag since traffic to and from .onion addresses are already
routed via Tor due to the `--onion` flag.

<a name="BridgeCLIExample" />

**4.2 Command Line Example**<br />

```bash
$ ./hcd --onion=127.0.0.1:9050 --externalip=fooanon.onion
```

<a name="BridgeConfigFileExample" />

**4.3 Config File Example**<br />

```text
[Application Options]

onion=127.0.0.1:9050
externalip=fooanon.onion
```

<a name="TorStreamIsolation" />

### 5. Tor Stream Isolation

<a name="TorStreamIsolationDescription" />

**5.1 Description**<br />

Tor stream isolation forces Tor to build a new circuit for each connection
making it harder to correlate connections.

hcd provides support for Tor stream isolation by using the `--torisolation`
flag.  This option requires --proxy or --onionproxy to be set.

<a name="TorStreamIsolationCLIExample" />

**5.2 Command Line Example**<br />

```bash
$ ./hcd --proxy=127.0.0.1:9050 --torisolation
```

<a name="TorStreamIsolationFileExample" />

**5.3 Config File Example**<br />

```text
[Application Options]

proxy=127.0.0.1:9050
torisolation=1
```


docs/configure_rpc_server_listen_interfaces.md
hcd allows you to bind the RPC server to specific interfaces which enables you
to setup configurations with varying levels of complexity.  The `rpclisten`
parameter can be specified on the command line as shown below with the -- prefix
or in the configuration file without the -- prefix (as can all long command line
options).  The configuration file takes one entry per line.

A few things to note regarding the RPC server:
* The RPC server will **not** be enabled unless the `rpcuser` and `rpcpass`
  options are specified.
* When the `rpcuser` and `rpcpass` and/or `rpclimituser` and `rpclimitpass`
  options are specified, the RPC server will only listen on localhost IPv4 and
  IPv6 interfaces by default.  You will need to override the RPC listen
  interfaces to include external interfaces if you want to connect from a remote
  machine.
* The RPC server has TLS enabled by default, even for localhost.  You may use
  the `--notls` option to disable it, but only when all listeners are on
  localhost interfaces.
* The `--rpclisten` flag can be specified multiple times to listen on multiple
  interfaces as a couple of the examples below illustrate.
* The RPC server is disabled by default when using the `--regtest` and
  `--simnet` networks.  You can override this by specifying listen interfaces.

Command Line Examples:

|Flags|Comment|
|----------|------------|
|--rpclisten=|all interfaces on default port which is changed by `--testnet`|
|--rpclisten=0.0.0.0|all IPv4 interfaces on default port which is changed by `--testnet`|
|--rpclisten=::|all IPv6 interfaces on default port which is changed by `--testnet`|
|--rpclisten=:14009|all interfaces on port 14009|
|--rpclisten=0.0.0.0:14009|all IPv4 interfaces on port 14009|
|--rpclisten=[::]:14009|all IPv6 interfaces on port 14009|
|--rpclisten=127.0.0.1:14009|only IPv4 localhost on port 14009|
|--rpclisten=[::1]:14009|only IPv6 localhost on port 14009|
|--rpclisten=:8336|all interfaces on non-standard port 8336|
|--rpclisten=0.0.0.0:8336|all IPv4 interfaces on non-standard port 8336|
|--rpclisten=[::]:8336|all IPv6 interfaces on non-standard port 8336|
|--rpclisten=127.0.0.1:8337 --listen=[::1]:14009|IPv4 localhost on port 8337 and IPv6 localhost on port 14009|
|--rpclisten=:14009 --listen=:8337|all interfaces on ports 14009 and 8337|

The following config file would configure the hcd RPC server to listen to all interfaces on the default port, including external interfaces, for both IPv4 and IPv6:

```text
[Application Options]

rpclisten=
```


cmd/hcctl/config.go
// Copyright (c) 2013-2015 The btcsuite developers
// Copyright (c) 2015-2017 The Decred developers
// Copyright (c) 2018-2020 The Hc developers
// Use of this source code is governed by an ISC
// license that can be found in the LICENSE file.

package main

import (
	"fmt"
	"io/ioutil"
	"net"
	"os"
	"os/user"
	"path/filepath"
	"regexp"
	"runtime"
	"strings"

	"github.com/HcashOrg/hcd/hcjson"
	"github.com/HcashOrg/hcd/hcutil"

	flags "github.com/jessevdk/go-flags"
)

const (
	// unusableFlags are the command usage flags which this utility are not
	// able to use.  In particular it doesn't support websockets and
	// consequently notifications.
	unusableFlags = hcjson.UFWebsocketOnly | hcjson.UFNotification
)

var (
	hcdHomeDir             = hcutil.AppDataDir("hcd", false)
	hcctlHomeDir           = hcutil.AppDataDir("hcctl", false)
	hcwalletHomeDir        = hcutil.AppDataDir("hcwallet", false)
	defaultConfigFile      = filepath.Join(hcctlHomeDir, "hcctl.conf")
	defaultRPCServer       = "localhost"
	defaultWalletRPCServer = "localhost"
	defaultRPCCertFile     = filepath.Join(hcdHomeDir, "rpc.cert")
	defaultWalletCertFile  = filepath.Join(hcwalletHomeDir, "rpc.cert")
)

// listCommands categorizes and lists all of the usable commands along with
// their one-line usage.
func listCommands() {
	const (
		categoryChain uint8 = iota
		categoryWallet
		numCategories
	)

	// Get a list of registered commands and categorize and filter them.
	cmdMethods := hcjson.RegisteredCmdMethods()
	categorized := make([][]string, numCategories)
	for _, method := range cmdMethods {
		flags, err := hcjson.MethodUsageFlags(method)
		if err != nil {
			// This should never happen since the method was just
			// returned from the package, but be safe.
			continue
		}

		// Skip the commands that aren't usable from this utility.
		if flags&unusableFlags != 0 {
			continue
		}

		usage, err := hcjson.MethodUsageText(method)
		if err != nil {
			// This should never happen since the method was just
			// returned from the package, but be safe.
			continue
		}

		// Categorize the command based on the usage flags.
		category := categoryChain
		if flags&hcjson.UFWalletOnly != 0 {
			category = categoryWallet
		}
		categorized[category] = append(categorized[category], usage)
	}

	// Display the command according to their categories.
	categoryTitles := make([]string, numCategories)
	categoryTitles[categoryChain] = "Chain Server Commands:"
	categoryTitles[categoryWallet] = "Wallet Server Commands (--wallet):"
	for category := uint8(0); category < numCategories; category++ {
		fmt.Println(categoryTitles[category])
		for _, usage := range categorized[category] {
			fmt.Println(usage)
		}
		fmt.Println()
	}
}

// config defines the configuration options for hcctl.
//
// See loadConfig for details on the configuration load process.
type config struct {
	ShowVersion     bool   `short:"V" long:"version" description:"Display version information and exit"`
	ListCommands    bool   `short:"l" long:"listcommands" description:"List all of the supported commands and exit"`
	ConfigFile      string `short:"C" long:"configfile" description:"Path to configuration file"`
	RPCUser         string `short:"u" long:"rpcuser" description:"RPC username"`
	RPCPassword     string `short:"P" long:"rpcpass" default-mask:"-" description:"RPC password"`
	RPCServer       string `short:"s" long:"rpcserver" description:"RPC server to connect to"`
	WalletRPCServer string `short:"w" long:"walletrpcserver" description:"Wallet RPC server to connect to"`
	RPCCert         string `short:"c" long:"rpccert" description:"RPC server certificate chain for validation"`
	PrintJSON       bool   `short:"j" long:"json" description:"Print json messages sent and received"`
	NoTLS           bool   `long:"notls" description:"Disable TLS"`
	Proxy           string `long:"proxy" description:"Connect via SOCKS5 proxy (eg. 127.0.0.1:9050)"`
	ProxyUser       string `long:"proxyuser" description:"Username for proxy server"`
	ProxyPass       string `long:"proxypass" default-mask:"-" description:"Password for proxy server"`
	TestNet         bool   `long:"testnet" description:"Connect to testnet"`
	SimNet          bool   `long:"simnet" description:"Connect to the simulation test network"`
	TLSSkipVerify   bool   `long:"skipverify" description:"Do not verify tls certificates (not recommended!)"`
	Wallet          bool   `long:"wallet" description:"Connect to wallet"`
}

// normalizeAddress returns addr with the passed default port appended if
// there is not already a port specified.
func normalizeAddress(addr string, useTestNet, useSimNet, useWallet bool) string {
	_, _, err := net.SplitHostPort(addr)
	if err != nil {
		var defaultPort string
		switch {
		case useTestNet:
			if useWallet {
				defaultPort = "12010"
			} else {
				defaultPort = "12009"
			}
		case useSimNet:
			if useWallet {
				defaultPort = "13010"
			} else {
				defaultPort = "13009"
			}
		default:
			if useWallet {
				defaultPort = "14010"
			} else {
				defaultPort = "14009"
			}
		}

		return net.JoinHostPort(addr, defaultPort)
	}
	return addr
}

// cleanAndExpandPath expands environement variables and leading ~ in the
// passed path, cleans the result, and returns it.
func cleanAndExpandPath(path string) string {
	// NOTE: The os.ExpandEnv doesn't work with Windows cmd.exe-style
	// %VARIABLE%, but the variables can still be expanded via POSIX-style
	// $VARIABLE.
	path = os.ExpandEnv(path)

	if !strings.HasPrefix(path, "~") {
		return filepath.Clean(path)
	}

	// Expand initial ~ to the current user's home directory, or ~otheruser
	// to otheruser's home directory.  On Windows, both forward and backward
	// slashes can be used.
	path = path[1:]

	var pathSeparators string
	if runtime.GOOS == "windows" {
		pathSeparators = string(os.PathSeparator) + "/"
	} else {
		pathSeparators = string(os.PathSeparator)
	}

	userName := ""
	if i := strings.IndexAny(path, pathSeparators); i != -1 {
		userName = path[:i]
		path = path[i:]
	}

	homeDir := ""
	var u *user.User
	var err error
	if userName == "" {
		u, err = user.Current()
	} else {
		u, err = user.Lookup(userName)
	}
	if err == nil {
		homeDir = u.HomeDir
	}
	// Fallback to CWD if user lookup fails or user has no home directory.
	if homeDir == "" {
		homeDir = "."
	}

	return filepath.Join(homeDir, path)
}

// filesExists reports whether the named file or directory exists.
func fileExists(name string) bool {
	if _, err := os.Stat(name); err != nil {
		if os.IsNotExist(err) {
			return false
		}
	}
	return true
}

// loadConfig initializes and parses the config using a config file and command
// line options.
//
// The configuration proceeds as follows:
// 	1) Start with a default config with sane settings
// 	2) Pre-parse the command line to check for an alternative config file
// 	3) Load configuration file overwriting defaults with any specified options
// 	4) Parse CLI options and overwrite/add any specified options
//
// The above results in functioning properly without any config settings
// while still allowing the user to override settings with config files and
// command line options.  Command line options always take precedence.
func loadConfig() (*config, []string, error) {
	// Default config.
	cfg := config{
		ConfigFile:      defaultConfigFile,
		RPCServer:       defaultRPCServer,
		RPCCert:         defaultRPCCertFile,
		WalletRPCServer: defaultWalletRPCServer,
	}

	// Pre-parse the command line options to see if an alternative config
	// file, the version flag, or the list commands flag was specified.  Any
	// errors aside from the help message error can be ignored here since
	// they will be caught by the final parse below.
	preCfg := cfg
	preParser := flags.NewParser(&preCfg, flags.HelpFlag)
	_, err := preParser.Parse()
	if err != nil {
		if e, ok := err.(*flags.Error); ok && e.Type != flags.ErrHelp {
			fmt.Fprintln(os.Stderr, err)
			fmt.Fprintln(os.Stderr, "")
			fmt.Fprintln(os.Stderr, "The special parameter `-` "+
				"indicates that a parameter should be read "+
				"from the\nnext unread line from standard input.")
			os.Exit(1)
		} else if ok && e.Type == flags.ErrHelp {
			fmt.Fprintln(os.Stdout, err)
			fmt.Fprintln(os.Stdout, "")
			fmt.Fprintln(os.Stdout, "The special parameter `-` "+
				"indicates that a parameter should be read "+
				"from the\nnext unread line from standard input.")
			os.Exit(0)
		}
	}

	// Show the version and exit if the version flag was specified.
	appName := filepath.Base(os.Args[0])
	appName = strings.TrimSuffix(appName, filepath.Ext(appName))
	usageMessage := fmt.Sprintf("Use %s -h to show options", appName)
	if preCfg.ShowVersion {
		fmt.Println(appName, "version", version())
		os.Exit(0)
	}

	// Show the available commands and exit if the associated flag was
	// specified.
	if preCfg.ListCommands {
		listCommands()
		os.Exit(0)
	}

	if !fileExists(preCfg.ConfigFile) {
		err := createDefaultConfigFile(preCfg.ConfigFile)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error creating a default config file: %v\n", err)
		}
	}

	// Load additional config from file.
	parser := flags.NewParser(&cfg, flags.Default)
	err = flags.NewIniParser(parser).ParseFile(preCfg.ConfigFile)
	if err != nil {
		if _, ok := err.(*os.PathError); !ok {
			fmt.Fprintf(os.Stderr, "Error parsing config file: %v\n",
				err)
			fmt.Fprintln(os.Stderr, usageMessage)
			return nil, nil, err
		}
	}

	// Parse command line options again to ensure they take precedence.
	remainingArgs, err := parser.Parse()
	if err != nil {
		if e, ok := err.(*flags.Error); !ok || e.Type != flags.ErrHelp {
			fmt.Fprintln(os.Stderr, usageMessage)
		}
		return nil, nil, err
	}

	// Multiple networks can't be selected simultaneously.
	numNets := 0
	if cfg.TestNet {
		numNets++
	}
	if cfg.SimNet {
		numNets++
	}
	if numNets > 1 {
		str := "%s: the testnet and simnet params can't be used " +
			"together -- choose one of the two"
		err := fmt.Errorf(str, "loadConfig")
		fmt.Fprintln(os.Stderr, err)
		return nil, nil, err
	}

	// Override the RPC certificate if the --wallet flag was specified and
	// the user did not specify one.
	if cfg.Wallet && cfg.RPCCert == defaultRPCCertFile {
		cfg.RPCCert = defaultWalletCertFile
	}

	// When the --wallet flag is specified, use the walletrpcserver port
	// if specified.
	if cfg.Wallet && cfg.WalletRPCServer != defaultWalletRPCServer {
		cfg.RPCServer = cfg.WalletRPCServer
	}
	// Handle environment variable expansion in the RPC certificate path.
	cfg.RPCCert = cleanAndExpandPath(cfg.RPCCert)

	// Add default port to RPC server based on --testnet and --wallet flags
	// if needed.
	cfg.RPCServer = normalizeAddress(cfg.RPCServer, cfg.TestNet,
		cfg.SimNet, cfg.Wallet)

	return &cfg, remainingArgs, nil
}

// createDefaultConfig creates a basic config file at the given destination path.
// For this it tries to read the hcd config file at its default path, and extract
// the RPC user and password from it.
func createDefaultConfigFile(destinationPath string) error {
	// Nothing to do when there is no existing hcd conf file at the default
	// path to extract the details from.
	hcdConfigPath := filepath.Join(hcdHomeDir, "hcd.conf")
	if !fileExists(hcdConfigPath) {
		return nil
	}

	// Read hcd.conf from its default path
	hcdConfigFile, err := os.Open(hcdConfigPath)
	if err != nil {
		return err
	}
	defer hcdConfigFile.Close()
	content, err := ioutil.ReadAll(hcdConfigFile)
	if err != nil {
		return err
	}

	// Extract the rpcuser
	rpcUserRegexp, err := regexp.Compile(`(?m)^\s*rpcuser=([^\s]+)`)
	if err != nil {
		return err
	}
	userSubmatches := rpcUserRegexp.FindSubmatch(content)
	if userSubmatches == nil {
		// No user found, nothing to do
		return nil
	}

	// Extract the rpcpass
	rpcPassRegexp, err := regexp.Compile(`(?m)^\s*rpcpass=([^\s]+)`)
	if err != nil {
		return err
	}
	passSubmatches := rpcPassRegexp.FindSubmatch(content)
	if passSubmatches == nil {
		// No password found, nothing to do
		return nil
	}

	// Create the destination directory if it does not exists
	err = os.MkdirAll(filepath.Dir(destinationPath), 0700)
	if err != nil {
		return err
	}

	// Create the destination file and write the rpcuser and rpcpass to it
	dest, err := os.OpenFile(destinationPath,
		os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0600)
	if err != nil {
		return err
	}
	defer dest.Close()

	dest.WriteString(fmt.Sprintf("rpcuser=%s\nrpcpass=%s",
		string(userSubmatches[1]), string(passSubmatches[1])))

	return nil
}
